pkg/controller/registration/registration_controller.go

```diff
@@ -248,21 +248,12 @@ func registerDefaultController(r *ReconcileRegistration, config *config.Controll
 
 		// If we have a choice of controllers, construct predicates to choose between them
 		if hasDirectController && (hasTerraformController || hasDCLController) {
-			reconcileGate := registry.GetReconcileGate(gvk.GroupKind())
-			if reconcileGate != nil {
+			if reconcileGate := registry.GetReconcileGate(gvk.GroupKind()); reconcileGate != nil {
 				// If reconcile gate is enabled for this gvk, generate a controller-runtime predicate that will
 				// run the direct reconciler only when the reconcile gate returns true.
 				useDirectReconcilerPredicate = kccpredicate.NewReconcilePredicate(r.mgr.GetClient(), gvk, reconcileGate)
 				useLegacyPredicate = kccpredicate.NewInverseReconcilePredicate(r.mgr.GetClient(), gvk, reconcileGate)
-			}
-
-			if !hasTerraformController && !hasDCLController {
-				// We're always going to use the direct reconciler
-				useDirectReconcilerPredicate = nil
-				useLegacyPredicate = nil
-			}
-
-			if (hasTerraformController || hasDCLController) && useDirectReconcilerPredicate == nil {
+			} else {
 				logger.Error(fmt.Errorf("no predicate where we have multiple controllers"), "skipping direct controller registration", "group", gvk.Group, "version", gvk.Version, "kind", gvk.Kind)
 				hasDirectController = false
```

Comment:
Previously we decouple the CRDs and controllers and let the labels on CRD to choose the controller. It only moves from TF to direct but not vice versa (which is an important precondition).  

Now the controller decision is switched to the CR object. 

I like the reconcile gate idea overall (except that it is very challenge to figure out which reconciler is being used, that could make triaging more difficult). I think my concern is mainly that the reconcile gate has changed the concept about some basic concept about the direct registry and label based reconciliations, that is not covered by any of our guides, tools, and test.